File: test/resolution_factor_test/resolution_factor_test.cpp
#include <FQlib.h>

void get_quad_(const double *img, const int img_size, const double weight_sigma_sq, double &quad_size,int tag, char*inform)
{
    // calculate the gaussian-weighted quadrupole of a galaxy or PSF image 
    double temp_quad, temp_norm;
    int i,j,m;
    double ry_sq, r_sq;
    double cen, wei_coeff,wei_img;
    
    // the image center
    cen = img_size*0.5-0.5;

    wei_coeff = 0.5/weight_sigma_sq;

    temp_quad = 0;
    temp_norm = 0;
    for(i=0; i<img_size; i++)
    {   
        ry_sq = (i - cen)*(i-cen);
        m = i*img_size;
        for(j=0; j<img_size; j++)
        {
            r_sq = (j - cen)*(j-cen) + ry_sq;

            wei_img = exp(-r_sq*wei_coeff)*img[m+j];
   
            temp_quad += wei_img*r_sq;
            temp_norm += wei_img;
        }
    }
    tag = 1;
    if(temp_quad< 0 or temp_norm<=0)
    {
        sprintf(inform,"%.5f, %.5f",temp_quad, temp_norm);
        tag = -1;
    }
    
    quad_size = temp_quad/temp_norm;
}

int main(int argc, char **argv)
{
    char img_path[200], data_path[200], set_name[30];
    double *img, *stamp;
    double *sex_data, *data;
    int i,j,k,tag;
    int nx, ny, size;
    int sex_row, sex_col;

    double st1, st2;

    double gal_quad;
    double eff_radius_sq;

    int quad_tag;
    char inform[40];

    size = 64;
    nx = 100;
    ny = 100;

    img = new double[nx*ny*size*size];
    stamp = new double[size*size];

    sex_row = 10000000;
    sex_col = 4;
    sex_data = new double[sex_row*sex_col];

    data = new double[nx*ny]{};

    sprintf(img_path,"/mnt/ddnfs/data_users/hkli/selection_bias/paper_data/pts_dimmer/1/gal_chip_0000.fits");
    read_fits(img_path, img);

    sprintf(data_path, "/mnt/ddnfs/data_users/hkli/selection_bias/paper_data/pts_dimmer/result/data/sex2_1.5/sex_1.hdf5");
    sprintf(set_name,"/data");
    read_h5(data_path, set_name, sex_data);
    st1 = clock();

    for(i=0;i<ny;i++)
    {
        for(j=0;j<nx;j++)
        {
            tag = i*nx + j;
            segment(img, stamp, tag, size, nx, ny);
            // \pi * r^2 = pixel number
            eff_radius_sq = sex_data[tag*sex_col + 3]/Pi;           
    
            if (eff_radius_sq > 0)            
            {
                get_quad(stamp, size, eff_radius_sq, gal_quad);
                data[tag] = gal_quad;
            }

            if(tag < 20)
            {
                std::cout<<sex_data[tag*sex_col + 3]<<" "<<eff_radius_sq<<" "<<data[tag]<<std::endl;
            }

        }

    }

    sprintf(data_path,"quad_size.hdf5");
    sprintf(set_name,"/data");
    write_h5(data_path, set_name, data,nx*ny,1,true);
    st2 = clock();
    std::cout<<(st2-st1)/CLOCKS_PER_SEC<<std::endl;

    delete[] sex_data;
    delete[] data;
    delete[] img;
    delete[] stamp;

    return 0;
}

File: test/resolution_factor_test/resolution_factor_test.py
import os
my_home = os.popen("echo $HOME").readlines()[0][:-1]
from sys import path
path.append("E:/Github/astrophy-research/mylib/")
path.append('%s/work/mylib/' % my_home)
import h5py
import numpy
from Fourier_Quad import Fourier_Quad
import tool_box
from astropy.io import fits
import time
from plot_tool import Image_Plot


t1 = time.time()




size = 64
num = 10000
fq = Fourier_Quad(size, 123)

img = fits.open("/mnt/ddnfs/data_users/hkli/selection_bias/paper_data/pts_dimmer/1/gal_chip_0000.fits")[0].data
h5f_sex = h5py.File("/mnt/ddnfs/data_users/hkli/selection_bias/paper_data/pts_dimmer/result/data/sex2_1.5/sex_1.hdf5","r")
sex_data = h5f_sex["/data"].value
h5f_sex.close()

gals = fq.segment(img)

quads = numpy.zeros((num,))

for i in range(len(gals)):
    eff_radius = sex_data[i,3]/numpy.pi
    if eff_radius > 0:
        quads[i] = tool_box.get_quad(gals[i], eff_radius, size)
    # if i < 20:
    #     print(sex_data[i,3],eff_radius, quads[i])

h5f = h5py.File("quad_size.hdf5", "r")
quad_cpp = h5f["/data"].value[:,0]
h5f.close()

# img = Image_Plot()
# img.subplots(1,1)
# img.axs[0][0].hist(quads,100)
# img.save_img("quad_hist.png")

diff = quads - quad_cpp
t2 =time.time()

print(diff.max(), diff.min(), t2-t1)
